src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java

```diff
@@ -0,0 +1,412 @@
+                case POST_URL_ENC_UPDATE -> {
+                    request.uri(connection.endpoint())
+                            .header("Content-Type", "application/x-www-form-urlencoded")
+                            .POST(HttpRequest.BodyPublishers.ofString(
+                                    urlEncode(Collections.singletonList(
+                                            new String[]{"update" /* query is already URL encoded */,
+                                                    new String(queryStream.readAllBytes(), StandardCharsets.UTF_8)}))));
+                }
+                case POST_UPDATE -> {
+                    request.uri(connection.endpoint())
+                            .header("Content-Type", "application/sparql-update")
+                            .POST(HttpRequest.BodyPublishers.ofInputStream(streamSupplier.getStreamSupplier()));
+                }
+            }
+            return request.build();
+        }
+    }
+
+
+    public record Config(
+            Integer number,
+            QueryHandler queries,
+            CompletionTarget completionTarget,
+            ConnectionConfig connection,
+            Duration timeout,
+            String acceptHeader /* e.g. application/sparql-results+json */,
+            RequestFactory.RequestType requestType,
+            Boolean parseResults
+    ) implements HttpWorker.Config {
+        public Config(Integer number,
+                      @JsonProperty(required = true) QueryHandler queries,
+                      @JsonProperty(required = true) CompletionTarget completionTarget,
+                      @JsonProperty(required = true) ConnectionConfig connection,
+                      @JsonProperty(required = true) Duration timeout,
+                      String acceptHeader,
+                      RequestFactory.RequestType requestType,
+                      Boolean parseResults) {
+            this.number = number == null ? 1 : number;
+            this.queries = queries;
+            this.completionTarget = completionTarget;
+            this.connection = connection;
+            this.timeout = timeout;
+            this.acceptHeader = acceptHeader;
+            this.requestType = requestType == null ? RequestFactory.RequestType.GET_QUERY : requestType;
+            this.parseResults = parseResults == null || parseResults;
+        }
+    }
+
+    record HttpExecutionResult(
+            int queryID,
+            Optional<HttpResponse<InputStream>> response,
+            Instant requestStart,
+            Duration duration,
+            Optional<BigByteArrayOutputStream> outputStream,
+            OptionalLong actualContentLength,
+            OptionalLong hash,
+            Optional<Exception> exception
+    ) {
+        public boolean completed() {
+            return response.isPresent();
+        }
+
+        public boolean successful() {
+            if (completed() && exception.isEmpty())
+                return (response.get().statusCode() / 100) == 2;
```

Comment:
Verify that all unchecked optional accesses are safe. Please do research if there is any way to indicate that the check was done by other means.